File: frontend/src/lib/api/acquisition.ts
import { apiClient, extractErrorMessage } from "@/lib/api-client";

export interface AcquisitionSource {
  id: string;
  institution_id: string;
  source_url: string;
  source_name: string;
  source_type: string;
  description: string | null;
  data_status: string;
  data_confidence: number | null;
  is_active: boolean;
  is_demo: boolean;
  robots_allowed: boolean | null;
  created_at: string;
}

export interface AcquisitionJob {
  id: string;
  institution_id: string;
  status: string;
  documents_downloaded: number;
  errors_count: number;
  error_message: string | null;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
}

export interface AcquisitionLog {
  id: string;
  job_id: string;
  source_url: string;
  success: boolean;
  status_code: number | null;
  file_type: string | null;
  error_message: string | null;
  robots_blocked: boolean;
  created_at: string;
}

export interface DownloadedDocument {
  id: string;
  institution_id: string;
  source_url: string;
  title: string;
  file_type: string;
  document_type: string;
  data_status: string;
  checksum: string | null;
  is_duplicate: boolean;
  created_at: string;
  // Wave 3 extraction fields
  extraction_status?: string;
  meaningful_title?: string | null;
  title_source?: string | null;
}

export interface AcquisitionStatistics {
  institution_id: string | null;
  total_sources: number;
  active_sources: number;
  total_jobs: number;
  completed_jobs: number;
  failed_jobs: number;
  total_documents: number;
  total_errors: number;
  last_job_at: string | null;
}

function instParam(institutionId?: string): Record<string, string> {
  return institutionId ? { institution_id: institutionId } : {};
}

export const acquisitionApi = {
  getSources: async (institutionId?: string): Promise<AcquisitionSource[]> => {
    const res = await apiClient.get<AcquisitionSource[]>("/acquisition/sources", {
      params: instParam(institutionId),
    });
    return res.data;
  },
  getJobs: async (institutionId?: string): Promise<AcquisitionJob[]> => {
    const res = await apiClient.get<AcquisitionJob[]>("/acquisition/jobs", {
      params: instParam(institutionId),
    });
    return res.data;
  },
  getLogs: async (
    institutionId?: string,
    jobId?: string,
  ): Promise<AcquisitionLog[]> => {
    const params: Record<string, string> = { ...instParam(institutionId) };
    if (jobId) params.job_id = jobId;
    const res = await apiClient.get<AcquisitionLog[]>("/acquisition/logs", { params });
    return res.data;
  },
  getDownloads: async (institutionId?: string): Promise<DownloadedDocument[]> => {
    const res = await apiClient.get<DownloadedDocument[]>("/acquisition/downloads", {
      params: instParam(institutionId),
    });
    return res.data;
  },
  getStatistics: async (
    institutionId?: string,
  ): Promise<AcquisitionStatistics> => {
    const res = await apiClient.get<AcquisitionStatistics>(
      "/acquisition/statistics",
      { params: instParam(institutionId) },
    );
    return res.data;
  },
  startJob: async (
    institutionId: string,
    sourceIds?: string[],
  ): Promise<AcquisitionJob> => {
    try {
      const res = await apiClient.post<AcquisitionJob>("/acquisition/jobs/start", {
        institution_id: institutionId,
        source_ids: sourceIds ?? null,
      });
      return res.data;
    } catch (err) {
      throw new Error(extractErrorMessage(err));
    }
  },
  retryJob: async (jobId: string): Promise<AcquisitionJob> => {
    const res = await apiClient.post<AcquisitionJob>(`/acquisition/retry/${jobId}`);
    return res.data;
  },
};

File: backend/app/schemas/acquisition.py
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class AcquisitionSourceRead(BaseModel):
    id: uuid.UUID
    institution_id: uuid.UUID
    source_url: str
    source_name: str
    source_type: str
    description: str | None
    data_status: str
    data_confidence: float | None
    is_active: bool
    is_demo: bool
    robots_allowed: bool | None
    created_at: datetime

    class Config:
        from_attributes = True


class AcquisitionSourceCreate(BaseModel):
    institution_id: uuid.UUID
    source_url: str = Field(..., max_length=1000)
    source_name: str = Field(..., max_length=500)
    source_type: str = Field(default="official_website", max_length=100)
    description: str | None = None
    data_status: str = Field(default="needs_review")
    data_confidence: float | None = Field(default=None, ge=0.0, le=1.0)


class AcquisitionJobRead(BaseModel):
    id: uuid.UUID
    institution_id: uuid.UUID
    status: str
    documents_downloaded: int
    errors_count: int
    error_message: str | None
    started_at: datetime | None
    completed_at: datetime | None
    created_at: datetime

    class Config:
        from_attributes = True


class AcquisitionJobStart(BaseModel):
    institution_id: uuid.UUID
    source_ids: list[uuid.UUID] | None = None


class AcquisitionLogRead(BaseModel):
    id: uuid.UUID
    job_id: uuid.UUID
    source_url: str
    success: bool
    status_code: int | None
    file_type: str | None
    error_message: str | None
    robots_blocked: bool
    created_at: datetime

    class Config:
        from_attributes = True


class DownloadedDocumentRead(BaseModel):
    id: uuid.UUID
    institution_id: uuid.UUID
    source_url: str
    title: str
    file_type: str
    document_type: str
    data_status: str
    checksum: str | None
    is_duplicate: bool
    created_at: datetime

    class Config:
        from_attributes = True


class AcquisitionStatistics(BaseModel):
    institution_id: uuid.UUID | None
    total_sources: int
    active_sources: int
    total_jobs: int
    completed_jobs: int
    failed_jobs: int
    total_documents: int
    total_errors: int
    last_job_at: datetime | None
